### PR TITLE
native: add Dlaqr2

### PR DESCRIPTION
**Pull request by vladimir-ch**
The documentation of reference DLAQR2 is quite loose or even incorrect in some places, so most of the differences in this PR come from that. Further, some of the documented restrictions on input values are in fact allowed and handled in the reference, I'm stricter and usually panic. That's another source of tiny differences.

Also, it seems that having a set of matrices with known eigenvalues will be soon very useful / necessary for testing. Any suggestions on where to get them, and how and where to store them? Using random matrices and computing their eigenvalues in e.g. Octave is of course possible but I'm after "difficult" eigenvalue problems that would trigger some exceptional paths in the code not taken by random matrices. Matrix Market is an option, their matrices for eigenvalue problems seem to be all sparse which should not be a problem for moderate sizes. Any other options?

### CODE REVIEW

File: native/dlaqr2.go [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+//  kbot == n-1 or H[kbot+1,kbot] == 0,
+// otherwise Dlaqr2 will panic.
+//
+// nw is the deflation window size. It must hold that
+//  1 <= nw <= kbot-ktop+1,
+// otherwise Dlaqr2 will panic.
+//
+// iloz and ihiz specify the rows of the n×n matrix Z to which transformations
+// will be applied if wantz is true. It must hold that
+//  0 <= iloz <= ktop,  and  kbot <= ihiz < n,
```
Comment (vladimir-ch) [inline native/dlaqr2.go]: When n==0, ktop and kbot must have specific values. Is it worth documenting it?

Comment (kortschak) [inline native/dlaqr2.go]: Yes, I think so.

File: native/dlaqr2.go [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+import (
+	"math"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+)
+
+// Dlaqr2 performs the orthogonal similarity transformation of a n×n upper
```
Comment (kortschak) [inline native/dlaqr2.go]: s/a/an/

File: native/dlaqr2.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+// Hessenberg matrix to detect and deflate fully converged eigenvalues from a
+// trailing principal submatrix using aggressive early deflation [1].
+//
+// On return, H will be overwritten by a new Hessenberg matrix that is a
+// perturbation of an orthogonal similarity transformation of H. It is to be
```
Comment (kortschak) [inline native/dlaqr2.go]: s/to be//

File: native/dlaqr2.go [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+// hoped that on output H will have many zero subdiagonal entries.
+//
+// If wantt is true, the matrix H will be fully updated so that the
+// quasi-triangular Schur factor can be computed. If wantt is false, then only
+// enough of H will be updated to preserve the eigenvalues.
+//
+// If wantz is true, the orthogonal similarity transformation will be
+// accumulated into Z[iloz:ihiz+1,ktop:kbot+1], otherwise Z is not referenced.
+//
+// ktop and kbot determine an block [ktop:kbot+1,ktop:kbot+1] along the diagonal
```
Comment (kortschak) [inline native/dlaqr2.go]: s/an/a/

File: native/dlaqr2.go [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+// available for workspace. It must hold that nh <= nw, otherwise Dlaqr2 will
+// panic.
+//
+// wv and ldwv represent an nw×nw work matrix W. nv is the number of rows of W
+// available for workspace. It must hold that nv <= nw, otherwise Dlaqr2 will
+// panic.
+//
+// work must have length at least lwork and lwork must be at least max(1,2*nw),
+// otherwise Dlaqr2 will panic. Larger values of lwork may result in greater
+// efficiency. On return, work[0] will contain the optimal value of lwork.
```
Comment (kortschak) [inline native/dlaqr2.go]: Is this always, or only when lwork == -1?

Comment (kortschak) [inline native/dlaqr2.go]: Yes, it does. Why? Do they do this anywhere else? I don't recall this behaviour for any other function.

**Comment by kortschak** [comment]
Another option for finding useful test matrices would be to use dvyukov's fuzzer. @btracey has used this AFAIK.

File: native/dlaqr2.go [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+	}
+	if jw >= 3 {
+		t[(jw-1)*ldt+jw-3] = 0
+	}
+
+	ns = jw
+	ilst := infqr
+	// Deflation detection loop.
+	for ilst < ns {
+		var bulge bool
```
Comment (kortschak) [inline native/dlaqr2.go]: `bulge := false`

File: native/dlaqr2.go [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+		if ns >= 2 {
+			bulge = t[(ns-1)*ldt+ns-2] != 0
+		}
+		if !bulge {
+			// Real eigenvalue.
+			abst := math.Abs(t[(ns-1)*ldt+ns-1])
```
Comment (kortschak) [inline native/dlaqr2.go]: They actually call this `foo`‽

File: native/dlaqr2.go [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+
+	ns = jw
+	ilst := infqr
+	// Deflation detection loop.
+	for ilst < ns {
+		var bulge bool
+		if ns >= 2 {
+			bulge = t[(ns-1)*ldt+ns-2] != 0
+		}
+		if !bulge {
```
Comment (kortschak) [inline native/dlaqr2.go]: The code duplication here is troubling, but I don't see a way to factor it cleanly. I guess that's why it's like this.

File: native/dlaqr2.go [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+			work[0] = 1
+			impl.Dlaset(blas.Lower, jw-2, jw-2, 0, 0, t[2*ldt:], ldt)
+			impl.Dlarf(blas.Left, ns, jw, work[:ns], 1, tau, t, ldt, work[jw:])
+			impl.Dlarf(blas.Right, ns, ns, work[:ns], 1, tau, t, ldt, work[jw:])
+			impl.Dlarf(blas.Right, jw, ns, work[:ns], 1, tau, v, ldv, work[jw:])
+			impl.Dgehrd(jw, 0, ns-1, t, ldt, work[:jw-1], work[jw:], lwork-jw)
+		}
+
+		// Copy updated reduced window into place.
+		if kwtop >= 1 {
```
Comment (kortschak) [inline native/dlaqr2.go]: `> 0`? I assume this is a 1->0 indexing transformation.

File: native/dlaqr2.go [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+		}
+		impl.Dlacpy(blas.Upper, jw, jw, t, ldt, h[kwtop*ldh+kwtop:], ldh)
+		bi.Dcopy(jw-1, t[ldt:], ldt+1, h[(kwtop+1)*ldh+kwtop:], ldh+1)
+
+		// Accumulate orthogonal matrix in order to update H and Z, if
+		// requested.
+		if ns > 1 && s != 0 {
+			// work[:ns-1] contains the elementary reflectors stored
+			// by a call to Dgehrd above.
+			impl.Dormhr(blas.Right, blas.NoTrans, jw, ns, 0, ns-1, t, ldt, work[:ns-1], v, ldv, work[jw:], lwork-jw)
```
Comment (kortschak) [inline native/dlaqr2.go]: Split this line.

File: native/dlaqr2.go [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+					k = i + 2
+				}
+			}
+		}
+	}
+
+	// Restore shift/eigenvalue array from T.
+	for i := jw - 1; i >= infqr; {
+		if i == infqr || t[i*ldt+i-1] == 0 {
+			sr[kwtop+i] = t[i*ldt+i]
```
Comment (kortschak) [inline native/dlaqr2.go]: The Fortran here is bizarre.

File: native/dlaqr2.go (hunk 5 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: Yes, they do (not sure why). Inexhaustive search gave me at least Dgehrd, Dorghr, Dormhr, Dorgbr.

File: native/dlaqr2.go (hunk 7 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: :-)

File: native/dlaqr2.go (hunk 1 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: Done.

File: native/dlaqr2.go (hunk 2 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: Done.

File: native/dlaqr2.go (hunk 3 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: Done.

File: native/dlaqr2.go (hunk 4 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: Done.

File: native/dlaqr2.go (hunk 6 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: Done.

File: native/dlaqr2.go (hunk 9 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: I wanted to include the border value of kwtop in the condition but with `> 0` it seems better. Done.

File: native/dlaqr2.go (hunk 10 above)
Comment (vladimir-ch) [inline native/dlaqr2.go]: Done.

**Comment by kortschak** [comment]
Implementation LGTM, leaving tests for @btracey

**Comment by vladimir-ch** [comment]
Reading through DLAQR4 I've noticed that from two possible interpretations of DLAQR2's documentation about the matrices T and WV I chose the wrong one. I've pushed a commit that should fix it. PTAL @kortschak , just in case.

**Comment by vladimir-ch** [comment]
I've pushed a commit to check input slices in Dlaqr2 only after a workspace query is handled so that it is possible to pass all slices as nil. A workspace query should depend only on the input parameters, the matrices and vectors must not be accessed anyway or may not be available at the call site. Also, often the reference LAPACK uses arbitrary placeholders for arrays and relies on the fact that no checking is done (as is the case of workspace query to DLAQR2 in DLAQR4).

I also updated the test which in turn forced me to make the same change in Dgehrd and Dormhr. I'm sorry for such extensive change but I think it is for a good end.

**Comment by kortschak** [comment]
>Moreover, often the reference LAPACK uses arbitrary placeholders for arrays and relies on the
fact that no checking is done.

Yes, this practice gives me the heebie-jeebies.

**Comment by kortschak** [comment]
I think this last change is a nice one; it makes the intention of the code clearer (maybe NETLIB people could take the opportunity to have a named NULL matrix that is 1x1 and could be used as a place holder to the same effect).

LGTM, but again wait for @btracey.

**Comment by btracey** [comment]
Test LGTM. I didn't check the implementation.

The eigenvalue tests are generally hard to do, because most randomly generated matrices are ill-conditioned. Maybe one could (in a combined test?) generate random eigenvalues and check that they look good to me.

**Pull request merged**